Retry fixing blocked UI on client start.

ntoll opened the pull request:
# Description

Fixes #846 and replaces #870.

While rebasing #870 with current master I got into all sorts of difficulties. As a result I abandoned the work and re-created the work from #870 but based from current master. I also fixed a couple of issues to do with deleting sources.

Please see the conversation on #870 for the discussion of the approach of this fix. This PR simply exists to avoid rebase hell.

# Test Plan

The unit tests have been copied over from the original PR. I've also removed some source deletion based unit tests which no longer worked because the approach to deletion is completely different. Please check the tests are comprehensive and exercise the code in the correct way.

I also tested this branch with 1000 sources and it works as expected.

# Checklist

If these changes modify code paths involving cryptography, the opening of files in VMs or network (via the RPC service) traffic, Qubes testing in the staging environment is required. For fine tuning of the graphical user interface, testing in any environment in Qubes is required. Please check as applicable:

 - [ ] I have tested these changes in the appropriate Qubes environment
 - [ ] I do not have an appropriate Qubes OS workstation set up (the reviewer will need to test these changes)
 - [X] These changes should not need testing in Qubes

If these changes add or remove files other than client code, packaging logic (e.g., the AppArmor profile) may need to be updated. Please check as applicable:

 - [ ] I have submitted a separate PR to the [packaging repo](https://github.com/freedomofpress/securedrop-debian-packaging)
 - [X] No update to the packaging logic (e.g., AppArmor profile) is required for these changes
 - [ ] I don't know and would appreciate guidance

creviera commented on securedrop_client/gui/widgets.py:
```diff
@@ -910,50 +907,76 @@ def setup(self, controller):
+        self.add_source(existing_source_widgets, sources, current_source_id,
+                        has_source_selected)
+
+    def add_source(self, existing_source_widgets, sources, current_source_id,
```
I think we should keep this method and use it with some modifications. I see how this improves performance when there are a large number of sources to add at once, but in order to also support not recreating all source widgets every time we sync (which is a huge performance hit for every other situation, especially when there are many existing source widgets) I think you can just update the signature to something like:

```
def add_source(self, sources_to_add, slice_size=1)
```

There's no reason to keep track of the `current_source_id` or `has_source_selected` or `existing_source_widgets` since we are just adding new source widgets and none of them will be selected by default. Existing source widgets should be irrelevant since we are just adding new source widgets.

After upating the signature, you can add back all the changes to the `update` function and pass the list of sources where the uuid is not in the list of widget_uuids (just the sources that need to be added). This is one way to support both optimizations: (1) scheduling when there many source widgets to add, (2) no longer recreating existing source widgets every sync. Another way would be to only run this method on the first run of the client.

Since we are syncing up about client architecture today, I'll provide relevant notes if there are any.

creviera replied:
No relevant updates from our architecture discussion not already mentioned in the comment above.

ntoll replied:
OK... I'm working on this. To be clear, here's what I'm doing:

* Simplifying the code I've already done (see function signature).
* Only call this on application start.
* Any further source updates are as they currently are.
* (Also, rebase). :-)

That's it.

creviera replied:
That sounds good to me, although technically it's possible to hit a scenario where the client needs to download a number of sources that blocks the UI (does this start at 1000?) after application start: (1) the client starts in offline mode, (2) the user logs in and the client begins downloading 1000 sources, or (1) user logs in, (2) server is restarted because of a technical issue causing the client to be unable to download sources, (3) once server restarts the client starts downloading 1000 sources.  

If it's not too much work, we could cover more situations if you call your source add function when we add new sources from a sync.

creviera replied:
Looks like you addressed my comment by running the initial update if there are no sources in the source list, which covers both the offline case and the error-during-source-download case. I'll mark this as resolved.

ntoll commented:
@creviera :+1: on re-using this code but with modification given the more recent changes that have happened on master.

creviera commented:
@ntoll i think this change looks good after you respond to https://github.com/freedomofpress/securedrop-client/pull/944#discussion_r393835241

ntoll commented:
@creviera / @redshiftzero please never make me rebase anything like this again. :stuck_out_tongue: Please review... I have:

* [x] simplified the recursive function signature
* [x] ensured the chunked method of populating the source list only happens if there are NO sources in the source list. For all other states, the current behaviour is retained.
* [x] updated various tests to reflect the above changes.
* [x] screamed, "GIT! Why GIT, why...?" on several occasions as I tried to unpick / rebase / check / test and generally get frustrated with computers.

That's it..!

creviera commented on securedrop_client/gui/widgets.py:
```diff
@@ -903,6 +909,8 @@ class SourceList(QListWidget):
     }
     '''
 
+    delete_source_by_uuid = pyqtSignal(str)
```
where is this emitted?

creviera replied:
(if this isn't emitted anywhere then we should delete this)

creviera commented on securedrop_client/gui/widgets.py:
```diff
@@ -977,6 +985,44 @@ def update(self, sources: List[Source]) -> List[str]:
+        """
+        Initialise the list with the passed in list of sources.
+        """
+        sources.reverse()
```
you can avoid having to reverse the sources list by inserting new widgets to the top of the SourceList in your add_source method

redshiftzero commented on securedrop_client/gui/widgets.py:
```diff
@@ -709,10 +710,15 @@ def show_sources(self, sources: List[Source]):
+                # Then call the function to remove the wrapper and its children.
+                self.delete_conversation(source_uuid)
+        else:
+            # We have an empty source list, so do an initial update.
```
Agreed, let's file a followup for this improvement

creviera commented:
I did a first pass at the code and am moving on to the test plan to test with 1000 sources, and will track benchmark results as I go, and then can follow up with a more thorough code review.

redshiftzero commented:
added a commit with requested changes, let's now test this thoroughly ensuring that:

- [ ] ordering of the source list matches the JI
- [ ] there are no other regressions

btw I think there are other scalability issues with `set_snippet` which we'll need to investigate (existing on `master`).

creviera commented on securedrop_client/logic.py:
```diff
@@ -520,8 +520,6 @@ def update_sources(self):
         """
         sources = list(storage.get_local_sources(self.session))
-        if sources:
-            sources.sort(key=lambda x: x.last_updated)
```
I wonder if no longer sorting by `last_updated` improves performance? If the server already provides the list in `last_updated` order and we always insert the sources in the local database in the same order then we can assume that `get_local_sources` provides the source list in `last_updated` order. Is that your thinking?

redshiftzero replied:
I modified `get_local_sources` to provide the list in `last_updated` order in SQL (other places we use `get_local_sources` the ordering doesn't matter).

creviera replied:
i see the change now. thanks for the explanation.

creviera commented:
Okay so I'm going to create at least two followup issues: (1) for the sync timeout issue I mentioned in my test results, and (2) for the issue when old widgets haven't been removed yet from the source list and we end up using a slower add method.

I'm still looking into why it takes ~53 seconds before we see the source list populated.

creviera commented:
> I'm still looking into why it takes ~53 seconds before we see the source list populated.

The plan is to address the slowness issue (issue still needs to be created I think) outside of this PR. Looks like the failing test opens up two client instances when you run it by itself, which is not happening on master, so looking into why this is happening.

creviera commented:
Update: Turns out he multiple client windows is a result of when a test marked as flaky fails and runs again.

ntoll commented:
Folks... many many thanks for the comprehensive testing and checking. I really appreciate it. :+1: Here's hoping nobody ever has 1000s of sources..! :wink: 

FWIW, perhaps in future, when folks are working on the same piece of code, it'd be good to coordinate in some way (stacked branches or small incremental changes done in a clearly defined order with clearly defined roles for who covers what aspect of the work..?) so we don't end up with the rebase hell I found myself in yesterday (it took several attempts to get it right, and as @creviera noticed, there were still some parts of the code I'd missed to remove). Having said that, our difference in time-zones and remote working didn't help the situation, and I think we're in a relatively good place considering where we could have been.

Stonking stuff..! :rocket:
